clippy_lints/src/loops/manual_memcpy.rs

```diff
@@ -446,3 +450,36 @@ fn get_loop_counters<'a, 'tcx>(
             .into()
     })
 }
+
+fn is_array_length_equal_to_range(cx: &LateContext<'_>, start: &Expr<'_>, end: &Expr<'_>, arr: &Expr<'_>) -> bool {
+    fn extract_lit_value(expr: &Expr<'_>) -> Option<u128> {
+        if_chain! {
+            if let ExprKind::Lit(lit) = expr.kind;
+            if let ast::LitKind::Int(value, _) = lit.node;
+            then {
+                Some(value)
+            } else {
+                None
+            }
+        }
+    }
+
+    let arr_ty = cx.typeck_results().expr_ty(arr).peel_refs();
+
+    if let ty::Array(_, s) = arr_ty.kind() {
```

Comment:
Is there any case in which this isn't an array? If so, we should make a first matching on `arr` to see if it's kind `ExprKind::Array` before querying the type (which is a more expensive function)

Reply:
I think the `arr` type could be a `std::vec::Vec<i32, std::alloc::Global>` or other types.
However, I'm actually not sure about how to classify `arr` type before querying it, as `src.base.kind` or `dst.base.kind` is `ExprKind::Path`

Reply:
The indexing operation is its own kind, [`ExprKind::Index`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.ExprKind.html#variant.Index). Maybe there's something there?

If you cannot find any `ExprKind::Index` expression (which should be either `src` or `dst` without `.base`) it's fine, it can be accepted with a little performance oversight (but I'd really like for it to be pre-matched).

Reply:
I couldn't find any expressions of type `ExprKind::Index` in `src` or `dst`. 
I think the `src` and `dst` refer to `base_right` and `base_left` respectively, as provided in https://github.com/rust-lang/rust-clippy/blob/4e85e41e58832d1b5b8379dd508b4908e5562d2b/clippy_lints/src/loops/manual_memcpy.rs#L63-L66

and these are analyzed as elements similar to slices.

Reply:
Okis, don't worry then.